File: excel-sheet-column-number/v2.py
class Solution:
    def titleToNumber(self, columnTitle: str | list[str]) -> int:
        num = 0
        for i, c in enumerate(columnTitle[::-1]):
            val = ord(c) - ord("A") + 1
            num += val * 26**i

        return num


# Runtime: 55 ms, faster than 29.73% of Python3 online submissions for Excel Sheet Column Number.
# Memory Usage: 13.8 MB, less than 98.67% of Python3 online submissions for Excel Sheet Column Number.

File: longest-palindromic-substring/v2.go
package v2

type State struct {
	Input                                string
	LongestStart, LongestEnd, LongestLen int
}

func (s *State) CheckPalindrome(startIdx, endIdx int) {
	if s.IsPalindrome(startIdx, endIdx) {
		thisLen := endIdx - startIdx + 1
		if thisLen > s.LongestLen {
			s.LongestLen = thisLen
			s.LongestStart = startIdx
			s.LongestEnd = endIdx
		}
	}
}

func (s *State) IsPalindrome(startIdx, endIdx int) bool {
	if startIdx >= endIdx {
		return true
	}
	if s.Input[startIdx] != s.Input[endIdx] {
		return false
	}
	return s.IsPalindrome(startIdx+1, endIdx-1)
}

func longestPalindrome(s string) string {
	state := State{
		Input: s,
	}

	for startIdx := range s {
		for endIdx := startIdx + 1; endIdx < len(s); endIdx++ {
			state.CheckPalindrome(startIdx, endIdx)
		}
	}

	return s[state.LongestStart : state.LongestEnd+1]
}

// Runtime: 2660 ms, faster than 5.04% of Go online submissions for Longest Palindromic Substring.
// Memory Usage: 2.7 MB, less than 38.59% of Go online submissions for Longest Palindromic Substring.
// oof...
